---
title: "R Notebook"
output: html_notebook
---

```{r}
# install cicero before running this script
rm(list=ls())
library(cicero)
```
# 0. Prepare data
 In this tutorial we use scAtac-seq data in 10x genomics database.
 You do not need to download these data when you analyze your scATAC data. 
```{r}
# create folder to store data
dir.create("data")

# download demo dataset from 10x genomics 
system("wget -O data/matrix.tar.gz http://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_filtered_peak_bc_matrix.tar.gz")

# unzip data
system("tar -xvf data/matrix.tar.gz -C data")
```
```{r}
# Just replace the data path below with the data path of your scATAC data if you have your scATAC data.

data_folder <- "data/filtered_peak_bc_matrix"
output_folder <- "cicero_output"
dir.create(output_folder)
```


# 1. Load data and make Cell Data Set (CDS) object 

## 1.1. Process data to make CDS object
```{r}
# read in matrix data using the Matrix package
indata <- Matrix::readMM(paste0(data_folder, "/matrix.mtx")) 
# binarize the matrix
indata@x[indata@x > 0] <- 1

# format cell info
cellinfo <- read.table(paste0(data_folder, "/barcodes.tsv"))
row.names(cellinfo) <- cellinfo$V1
names(cellinfo) <- "cells"

# format peak info
peakinfo <- read.table(paste0(data_folder, "/peaks.bed"))
names(peakinfo) <- c("chr", "bp1", "bp2")
peakinfo$site_name <- paste(peakinfo$chr, peakinfo$bp1, peakinfo$bp2, sep="_")
row.names(peakinfo) <- peakinfo$site_name

row.names(indata) <- row.names(peakinfo)
colnames(indata) <- row.names(cellinfo)

# make CDS
input_cds <-  suppressWarnings(newCellDataSet(indata,
                            phenoData = methods::new("AnnotatedDataFrame", data = cellinfo),
                            featureData = methods::new("AnnotatedDataFrame", data = peakinfo),
                            expressionFamily=VGAM::binomialff(),
                            lowerDetectionLimit=0))
input_cds@expressionFamily@vfamily <- "binomialff"
input_cds <- monocle::detectGenes(input_cds)

#Ensure there are no peaks included with zero reads
input_cds <- input_cds[Matrix::rowSums(exprs(input_cds)) >= 100,] 
```

# 2. QC and Filtering
```{r}
# Visualize peak_count_per_cell
hist(Matrix::colSums(exprs(input_cds)))


```
```{r}
# filter cells by peak_count
max_count <-  15000
min_count <- 2000
input_cds <- input_cds[,Matrix::colSums(exprs(input_cds)) >= min_count] 
input_cds <- input_cds[,Matrix::colSums(exprs(input_cds)) <= max_count] 

```



# 3. Process cicero-CDS object
```{r}
# Run cicero to get cis-regulatory networks
set.seed(2017)
input_cds <- detectGenes(input_cds)
input_cds <- estimateSizeFactors(input_cds)

input_cds <- reduceDimension(input_cds, max_components = 2, verbose=T,scaling = FALSE,relative_expr=FALSE,
                      reduction_method = 'tSNE', norm_method = "none")

tsne_coords <- t(reducedDimA(input_cds))
row.names(tsne_coords) <- row.names(pData(input_cds))
cicero_cds <- make_cicero_cds(input_cds, reduced_coordinates = tsne_coords)

saveRDS(cicero_cds, paste0(output_folder, "/cicero_cds.Rds"))

```

# 4. Run cicero to get cis-regulatory connection scores
```{r}
# import genome length, which is needed for the function, run_cicero 
mm10_chromosome_length <- read.table("./mm10_chromosome_length.txt")

# runc the main function
conns <- run_cicero(cicero_cds, mm10_chromosome_length) # Takes a few minutes to run

# save results
saveRDS(conns, paste0(output_folder, "/cicero_connections.Rds"))

# check results
head(conns)
```

```{r}
all_peaks <- row.names(exprs(input_cds))
write.csv(x = all_peaks, file = paste0(output_folder, "/all_peaks.csv"))
write.csv(x = conns, file = paste0(output_folder, "/cicero_connections.csv"))

```


